src/recv.rs

```diff
@@ -77,26 +76,32 @@ impl<const N: usize> ReceiveBuffer<N> {
     /// Panics if `data.len()` is greater than the amount of available bytes in the buffer and the
     /// data for `seq_num` has not already been written.
     pub fn write(&mut self, data: &[u8], seq_num: u16) {
-        if self.was_written(seq_num) {
+        // Reject packets that are either already received or too old.
+        if self.was_written(seq_num) || !seq_after(seq_num, self.ack_num()) {
             return;
         }
 
-        // TODO: Return error instead of panicking.
         if data.len() > self.available() {
             panic!("insufficient space in buffer");
         }
 
-        // Read all sequential data from pending packets.
+        // Clean up any pending packets that are now outdated.
+        // They should be strictly after ack_num()+1.
+        let current_ack_plus1 = self.ack_num().wrapping_add(1);
+        self.pending
+            .retain(|&seq, _| seq_after(seq, current_ack_plus1));
+
+        // Insert the new packet.
         self.pending.insert(seq_num, data.to_vec());
-        let start = self.init_seq_num.wrapping_add(1);
-        let mut next = start.wrapping_add(self.consumed);
+
+        // Process contiguous pending packets.
+        let mut next = self.ack_num().wrapping_add(1);
         while let Some(data) = self.pending.remove(&next) {
             let end = self.offset + data.len();
             self.buf.as_mut_slice()[self.offset..end].copy_from_slice(&data[..]);
-
             self.offset = end;
             self.consumed += 1;
-            next = next.wrapping_add(1);
+            next = self.ack_num().wrapping_add(1);
```

Comment:
```suggestion
            next = next.wrapping_add(1);
```

shouldn't this be next?

Reply:
No, because after processing a contiguous packet, it updates `self.consumed` (which in turn updates `ack_num()`). Instead of simply incrementing next, it recalculates it using `self.ack_num().wrapping_add(1)`. This ensures that any change to consumed is immediately reflected when determining the next expected packet. 

The fix makes the intention explicit: the next expected sequence is always `ack_num() + 1`. This can help prevent bugs in the future if the logic for updating contiguous data changes.